Use smart pointers in RemoteLayerTreeDrawingArea

pratiksha-99 opened the pull request:
#### 4f50b03bc5752991d16a637c470f499af8aa6b1b
<pre>
Use smart pointers in RemoteLayerTreeDrawingArea
<a href="https://bugs.webkit.org/show_bug.cgi?id=287386">https://bugs.webkit.org/show_bug.cgi?id=287386</a>

Reviewed by NOBODY (OOPS!).

Addressed the smart pointer static analyzer warnings.

* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
(WebCore::GraphicsLayerCA::protectedPlatformCALayer const):
* Source/WebKit/WebProcess/WebPage/DrawingArea.cpp:
(WebKit::DrawingArea::protectedWebPage const):
* Source/WebKit/WebProcess/WebPage/DrawingArea.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
(WebKit::RemoteLayerTreeDrawingArea::updateRenderingWithForcedRepaint):
(WebKit::RemoteLayerTreeDrawingArea::setViewExposedRect):
(WebKit::RemoteLayerTreeDrawingArea::exposedContentRect const):
(WebKit::RemoteLayerTreeDrawingArea::setExposedContentRect):
(WebKit::RemoteLayerTreeDrawingArea::updateRendering):
(WebKit::RemoteLayerTreeDrawingArea::didCompleteRenderingUpdateDisplayFlush):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4f50b03bc5752991d16a637c470f499af8aa6b1b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89776 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9305 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44725 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94773 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40549 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91828 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9692 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17582 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/69147 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26763 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92777 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7433 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81495 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49502 "Found 1 new API test failure: /WPE/TestSSL:/webkit/WebKitWebView/ephemeral-tls-errors (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/7155 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/35869 "Found 1 new test failure: http/tests/security/file-system-access-via-dataTransfer.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39682 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77502 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/36911 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96596 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16970 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/12456 "Found 3 new test failures: http/wpt/html/cross-origin-embedder-policy/require-corp.https.html imported/w3c/web-platform-tests/fullscreen/api/element-request-fullscreen-namespaces.html imported/w3c/web-platform-tests/html/browsers/the-window-object/open-close/open-features-tokenization-noreferrer.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/78014 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17221 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/77311 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77335 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21778 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/20384 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/10140 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16971 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/22295 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16712 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/20167 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/18503 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

RupinMittal commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
```diff
@@ -112,7 +112,8 @@
 {
-    m_remoteLayerTreeContext->gpuProcessConnectionWasDestroyed();
+    Ref remoteLayerTreeContext = m_remoteLayerTreeContext.get();
+    remoteLayerTreeContext->gpuProcessConnectionWasDestroyed();
```
This totally works! But it would be better to avoid creating the `remoteLayerTreeContext` local variable by using a protected function instead. I would create a new function `protectedRemoteLayerTreeContext()` and then do:
`protectedRemoteLayerTreeContext()->gpuProcessConnectionWasDestroyed();`

The same goes for the others like `m_webPage`, and `m_commitQueue`.

pratiksha-99 replied:
Thanks Rupin, I was in two minds earlier. I will work on it.

RupinMittal commented on Source/WebKit/WebProcess/WebPage/DrawingArea.h:
```diff
@@ -203,6 +203,7 @@ class DrawingArea : public RefCounted<DrawingArea>, public IPC::MessageReceiver,
     const DrawingAreaType m_type;
     DrawingAreaIdentifier m_identifier;
     WeakRef<WebPage> m_webPage;
+    Ref<WebPage> protectedWebPage() const;
```
Nit: As a general principle, we separate functions from variables rather than interleaving them. List all the functions first and then the variables. For example, I'd put `protectedWebPage` below `send`.

The same idea applies to `protectedConnection`, `protectedRemoteLayerTreeContext`, and `protectedCommitQueue`.

cdumez replied:
Yes, let's not fix member functions and data members.

RupinMittal reviewed (commented):
Nice! Just a small stylistic comment.

cdumez commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.h:
```diff
@@ -159,10 +159,12 @@ class RemoteLayerTreeDrawingArea : public DrawingArea, public WebCore::GraphicsL
         explicit BackingStoreFlusher(Ref<IPC::Connection>&&);
 
         Ref<IPC::Connection> m_connection;
+        Ref<IPC::Connection> protectedConnection() const { return m_connection.get(); }
```
ditto.

cdumez replied:
But also, can we make m_connection const to avoid a protected function?

cdumez commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.h:
```diff
@@ -159,10 +159,12 @@ class RemoteLayerTreeDrawingArea : public DrawingArea, public WebCore::GraphicsL
     };
 
     Ref<RemoteLayerTreeContext> m_remoteLayerTreeContext;
+    Ref<RemoteLayerTreeContext> protectedRemoteLayerTreeContext() const;
```
ditto.

cdumez replied:
But also, can we make `m_remoteLayerTreeContext` const to avoid a protected function?

cdumez commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.h:
```diff
@@ -190,6 +192,7 @@ class RemoteLayerTreeDrawingArea : public DrawingArea, public WebCore::GraphicsL
     bool m_deferredRenderingUpdateWhileWaitingForBackingStoreSwap { false };
 
     Ref<WorkQueue> m_commitQueue;
+    Ref<WorkQueue> protectedCommitQueue() const;
```
ditto.

cdumez replied:
But also, can we make `m_commitQueue` const to avoid requiring a protected function?

RupinMittal commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.h:
```diff
@@ -158,12 +158,12 @@ class RemoteLayerTreeDrawingArea : public DrawingArea, public WebCore::GraphicsL
         explicit BackingStoreFlusher(Ref<IPC::Connection>&&);
 
-        Ref<IPC::Connection> m_connection;
+        Ref<IPC::Connection> const m_connection;
```
Nit: WebKit convention for variables is to have the const first: `const Ref<IPC::Connection> m_connection;`

pratiksha-99 replied:
I will see if I can add this type of check to `Tools/Scripts/check-webkit-style` .

pratiksha-99 committed:
Use smart pointers in RemoteLayerTreeDrawingArea
https://bugs.webkit.org/show_bug.cgi?id=287386

Reviewed by NOBODY (OOPS!).

Addressed the smart pointer static analyzer warnings.

* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
(WebCore::GraphicsLayerCA::protectedPlatformCALayer const):
* Source/WebKit/WebProcess/WebPage/DrawingArea.cpp:
(WebKit::DrawingArea::protectedWebPage const):
* Source/WebKit/WebProcess/WebPage/DrawingArea.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
(WebKit::RemoteLayerTreeDrawingArea::updateRenderingWithForcedRepaint):
(WebKit::RemoteLayerTreeDrawingArea::setViewExposedRect):
(WebKit::RemoteLayerTreeDrawingArea::exposedContentRect const):
(WebKit::RemoteLayerTreeDrawingArea::setExposedContentRect):
(WebKit::RemoteLayerTreeDrawingArea::updateRendering):
(WebKit::RemoteLayerTreeDrawingArea::didCompleteRenderingUpdateDisplayFlush):
